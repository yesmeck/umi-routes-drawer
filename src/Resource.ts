import { castArray, without } from 'lodash';
import { singular } from 'pluralize';

export default class Resource {
  name: string;
  path: string;
  domain: string;
  only: string[];
  except: string[];

  constructor(name: string, options: any = {}) {
    this.name = name;
    this.path = options.path || this.name;
    this.domain = name;
    this.only = options.only ? castArray(options.only) : [];
    this.except = options.except ? castArray(options.except) : [];
  }

  actions() {
    return this.except.length > 0 ? without(this.availableActions(), ...this.except) : this.availableActions();
  }

  defaultActions() {
    return ['index', 'new', 'show', 'edit'];
  }

  availableActions() {
    return this.only.length > 0 ? this.only : this.defaultActions();
  }

  plural() {
    return this.name;
  }

  singular() {
    return singular(this.name);
  }

  collectionName() {
    return this.singular() == this.plural() ? '#{plural}_index' : this.plural();
  }

  memberName() {
    return this.singular();
  }

  resourceScope() {
    return this.domain;
  }

  collectionScope() {
    return this.path;
  }

  memberScope() {
    return `${this.path}/:id`;
  }

  newScope(newPath: string) {
    return `${this.path}/${newPath}`;
  }

  nestedParam() {
    return `${this.singular()}Id`;
  }

  nestedScope() {
    return `${this.path}/:${this.nestedParam()}`;
  }
}
