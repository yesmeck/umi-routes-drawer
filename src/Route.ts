export default class Route {
  name: string;
  path: string;
  precedence: number;

  constructor(name: string, path: string, precedence: number) {
    this.name = name;
    this.path = path;
    this.precedence = precedence;
  }
}
