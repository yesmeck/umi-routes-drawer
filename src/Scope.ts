export default class Scope {
  static OPTIONS = [
    'path',
    'shallow_path',
    'as',
    'shallow_prefix',
    'module',
    'controller',
    'action',
    'path_names',
    'shallow',
    'blocks',
    'defaults',
    'via',
    'format',
    'options',
    'to'
  ];

  static RESOURCE_SCOPES = ['resource', 'resources'];
  static RESOURCE_METHOD_SCOPES = ['collection', 'member', 'new'];

  hash: any;
  parent: Scope | null;
  scopeLevel: any;

  constructor(hash: any, parent: Scope | null = null, scopeLevel: string | null = null) {
    this.hash = hash;
    this.parent = parent;
    this.scopeLevel = scopeLevel;
  }

  new(hash: any) {
    return new Scope(hash, this, this.scopeLevel);
  }

  newLevel(level: string) {
    return new Scope(this.hash, this, level);
  }

  isResourceScope() {
    return Scope.RESOURCE_SCOPES.includes(this.scopeLevel);
  }

  isResourceMethodScope() {
    return Scope.RESOURCE_METHOD_SCOPES.includes(this.scopeLevel);
  }

  get(key: string) {
    const scope = this.find(node => node.hash[key]);
    return scope && scope.hash[key];
  }

  find(callback: (scope: Scope) => boolean) {
    let node: Scope | null = this;
    while (node !== null) {
      if (callback(node)) {
        return node;
      }
      node = node.parent;
    }
  }

  actionName(prefix: string, collectionName: string, memberName: string) {
    switch (this.scopeLevel) {
      case 'nested':
        return [prefix];
      case 'collection':
        return [prefix, collectionName];
      case 'new':
        return [prefix, 'new', memberName];
      case 'root':
        return [collectionName, prefix];
      default:
        return [memberName, prefix];
    }
  }

  options() {
    return Scope.OPTIONS;
  }
}
