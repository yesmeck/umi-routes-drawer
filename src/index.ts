import { IRoute } from 'umi-types';

type Block = () => void;

interface PageOptions {
  component: string;
  redirect?: string;
}

let routes: IRoute[] = [];
let currentRoutes = routes;

function reset() {
  routes = [];
  currentRoutes = routes;
}

export function draw(block: Block) {
  reset();
  block();
  return routes;
}

export function page(path: string, options: PageOptions) {
  currentRoutes.push({
    path,
    ...options
  });
}

export function resources(resources: string, options?: {}, block?: Block) {
  currentRoutes.push(
    {
      path: `/${resources}`,
      component: `/pages/${resources}/index`
    },
    {
      path: `/${resources}/new`,
      component: `/pages/${resources}/new`
    },
    {
      path: `/${resources}/:id`,
      component: `/pages/${resources}/show`
    },
    {
      page: `/${resources}/:id/edit`,
      component: `/pages/${resources}/edit`
    }
  );

  if (block) {
    block();
  }
}
