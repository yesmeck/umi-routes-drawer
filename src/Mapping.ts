import Scope from './Scope';
import RouteSet from './RouteSet';
import Route from './Route';

export default class Mapping {
  set: RouteSet;
  path: string;
  defaultDomain: string;
  defaultAction: string;

  static build(
    scope: Scope,
    set: RouteSet,
    path: string,
    domain: string,
    defaultAction: string,
    component: string,
    options: any
  ) {
    options = { ...scope.get('options'), ...options };
    return new Mapping(scope, set, path, domain, defaultAction, component, options);
  }

  constructor(
    scope: Scope,
    set: RouteSet,
    path: string,
    domain: string,
    defaultAction: string,
    component: string,
    options: any
  ) {
    this.set = set;
    this.path = path;
    this.defaultDomain = domain;
    this.defaultAction = defaultAction;
  }

  makeRoute(name: string, precedence: number) {
    return new Route(name, this.path, precedence);
  }
}
