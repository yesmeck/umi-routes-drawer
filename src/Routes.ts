import Mapping from './Mapping';
import Route from './Route';

export default class Routes {
  routes: Route[] = [];

  addRoute(name: string, mapping: Mapping) {
    const route = mapping.makeRoute(name, this.routes.length);
    this.routes.push(route);
    return route;
  }
}
