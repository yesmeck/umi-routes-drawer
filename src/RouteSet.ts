import Mapping from './Mapping';
import Routes from './Routes';
import Route from './Route';

export default class RouteSet {
  set: Routes;

  namedRoutes: { [name: string]: Route } = {};

  resourcesPathNames: any;

  constructor() {
    this.set = new Routes();
    this.resourcesPathNames = RouteSet.defaultResourcesPathNames();
  }

  static defaultResourcesPathNames() {
    return { new: 'new', edit: 'edit' };
  }

  addRoute(mapping: Mapping, name: string) {
    if (!name || !name.match(/^[_a-z]\w*$/i)) {
      throw new Error(`Invalid route name: '${name}'`);
    }

    if (name && this.namedRoutes[name]) {
      throw new Error(
        `Invalid route name, already in use: '${name}'
         You may have defined two routes with the same name using the \`:as\` option, or
         you may be overriding a route already defined by a resource with the same naming.
         For the latter, you can restrict the routes created with \`resources\` as explained here:
         "https://guides.rubyonrails.org/routing.html#restricting-the-routes-created`
      );
    }

    const route = this.set.addRoute(name, mapping);
    if (name) {
      this.namedRoutes[name] = route;
    }
  }

  toJSON() {
    return this.set.routes;
  }
}
