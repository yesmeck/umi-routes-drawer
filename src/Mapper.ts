import { compact, flatten, capitalize } from 'lodash';
import { Block } from './types';
import extractOptionsAndBlock from './helpers/extractOptionsAntBlock';
import Scope from './Scope';
import normalizePath from 'normalize-path';
import Resource from './Resource';
import RouteSet from './RouteSet';
import Mapping from './Mapping';

type Action = 'index' | 'new' | 'show' | 'edit';

interface ResourcesOptions {
  only?: Action[];
}

export default class Mapper {
  static CANONICAL_ACTIONS = ['index', 'new', 'show'];

  set: RouteSet;
  currentScope: Scope;

  constructor(set: RouteSet) {
    this.set = set;
    this.currentScope = new Scope({ pathNames: set.resourcesPathNames });
  }

  static normalizePath(path: string) {
    path = normalizePath(path);
    return path[0] === '/' ? path : `/${path}`;
  }

  static normalizeName(name: string) {
    return Mapper.normalizePath(name)
      .slice(1)
      .replace(/\//g, '_');
  }

  resources(...args: any) {
    const resources = args.slice();
    const [options, block] = extractOptionsAndBlock<ResourcesOptions>(resources);
    if (this.applyCommonBehaviorFor('resources', resources, options, block)) {
      return;
    }

    this.withScopeLevel('resources', () => {
      this.resourceScope(new Resource(resources.pop(), options), () => {
        block();

        this.collection(() => {
          if (
            this.parentResource()
              .actions()
              .includes('index')
          ) {
            this.page('index', {});
          }
        });

        if (
          this.parentResource()
            .actions()
            .includes('new')
        ) {
          this.new(() => {
            this.page('new', {});
          });
        }

        this.setMemberMappingsForResource();
      });
    });
  }

  new(block: Block) {
    if (!this.currentScope.isResourceScope()) {
      throw new Error("can't use new outside resource(s) scope");
    }

    this.withScopeLevel('new', () => {
      this.pathScope(this.parentResource().newScope(this.actionPath('new')), block);
    });
  }

  scope(...args: any) {
    const resources = args.slice();
    const [options, block] = extractOptionsAndBlock<any>(resources);
    const scope: any = {};

    if (args) {
      options.path = flatten(args).join('/');
    }

    if (options.only || options.except) {
      scope.actionOptions = { only: options.only, except: options.except };
    }

    let value;
    this.currentScope.options().forEach(option => {
      if (option === 'options') {
        value = options;
      } else {
        value = options.option || null;
      }
      if (value !== null) {
        // scope[option] = (this as any)[`merge${capitalize(option)}Scope`]((this.currentScope as any)[option], value);
      }
    });
    this.currentScope = this.currentScope.new(scope);
    block();
    this.currentScope = this.currentScope.parent!;
  }

  member(block: Block) {
    if (!this.currentScope.isResourceScope()) {
      throw new Error("can't use member outside resource(s) scope");
    }

    this.withScopeLevel('member', () => {
      this.pathScope(this.parentResource().memberScope(), block);
    });
  }

  collection(block: Block) {
    if (!this.currentScope.isResourceScope()) {
      throw new Error("can't use collection outside resource(s) scope");
    }

    this.withScopeLevel('collection', () => {
      this.pathScope(this.parentResource().collectionScope(), block);
    });
  }

  nested(block: Block) {
    if (!this.currentScope.isResourceScope()) {
      throw new Error("can't use nested outside resource(s) scope");
    }

    this.withScopeLevel('nested', () => {
      this.pathScope(this.parentResource().nestedScope(), () => {
        this.scope(this.nestedOptions(), block);
      });
    });
  }

  page(path: string, options: any = {}) {
    this.match(path, options);
  }

  match(path: string, options: any) {
    const paths = [path];
    this.mapMatch(paths, options);
  }

  mapMatch(paths: string[], options: any) {
    const domain = this.currentScope.get('domain');
    const component = options.component;
    paths.forEach(action => {
      const routeOptions = { ...options };
      this.decomposedMatch(action, domain, routeOptions, component);
    });
  }

  decomposedMatch(action: string, domain: string, options: any, component: string) {
    switch (this.currentScope.scopeLevel) {
      case 'resources':
        break;
      case 'resource':
        break;
      default:
        this.addRoute(action, domain, options, component);
        break;
    }
  }

  addRoute(action: string | null, domain: string, options: any, component: string) {
    let path = this.pathForAction(action!);
    if (!path) {
      throw new Error('path is required');
    }
    let defaultAction = options.action || this.currentScope.get('action');

    if (/^[\w\-\/]+$/.test(action!)) {
      if (!action!.includes('/')) {
        defaultAction = defaultAction || action!.replace(/-/g, '_');
      }
    } else {
      action = null;
    }

    const as = this.nameForAction(options.as, action!);

    path = Mapper.normalizePath(path);
    const mapping = Mapping.build(this.currentScope, this.set, path, domain, defaultAction, '', options);
    this.set.addRoute(mapping, as!);
  }

  isCanonicalAction(action: string) {
    return this.currentScope.isResourceMethodScope() && Mapper.CANONICAL_ACTIONS.includes(action);
  }

  nameForAction(as: string, action: string) {
    const prefix = this.prefixNameForAction(as, action);

    let collectionName: string;
    let memberName: string;
    if (this.parentResource()) {
      if (!action) {
        return null;
      }
      collectionName = this.parentResource().collectionName();
      memberName = this.parentResource().memberName();
    }
    const actionName = this.currentScope.actionName(prefix!, collectionName!, memberName!);
    const candidate = compact(actionName).join('_');
    return candidate;
  }

  prefixNameForAction(as: string, action: string) {
    let prefix;
    if (as) {
      prefix = as;
    } else if (!this.isCanonicalAction(action)) {
      prefix = action;
    }

    if (prefix && prefix != '/' && prefix) {
      return Mapper.normalizeName(prefix.replace(/-/g, '_'));
    }
  }

  pathForAction(action: string, path?: string) {
    if (path) {
      return `${this.currentScope.get('path')}/${path}`;
    }

    return this.isCanonicalAction(action)
      ? this.currentScope.get('path')
      : `${this.currentScope.get('path')}/${this.actionPath(action)}`;
  }

  actionPath(action: string) {
    return this.currentScope.get('pathNames')[action] || action;
  }

  call(block: Block<any>) {
    block({
      scope: this.scope.bind(this),
      page: this.page.bind(this),
      resources: this.resources.bind(this),
      member: this.member.bind(this),
      collection: this.collection.bind(this)
    });
  }

  withScopeLevel(kind: any, block: Block) {
    this.currentScope = this.currentScope.newLevel(kind);
    block();
    this.currentScope = this.currentScope.parent!;
  }

  resourceScope(resource: Resource, block: Block) {
    this.currentScope = this.currentScope.new({ scopeLevelResource: resource });
    this.domain(resource.resourceScope(), block);
    this.currentScope = this.currentScope.parent!;
  }

  pathScope(path: string, block: Block) {
    this.currentScope = this.currentScope.new({ path: this.mergePathScope(this.currentScope.get('path'), path) });
    block();
    this.currentScope = this.currentScope.parent!;
  }

  domain(domain: string, block: Block) {
    this.currentScope = this.currentScope.new({ domain });
    block();
    this.currentScope = this.currentScope.parent!;
  }

  setMemberMappingsForResource() {
    this.member(() => {
      if (
        this.parentResource()
          .actions()
          .includes('edit')
      ) {
        this.page('edit');
      }

      if (
        this.parentResource()
          .actions()
          .includes('show')
      ) {
        this.page('show');
      }
    });
  }

  nestedOptions() {
    return { as: this.parentResource().memberName };
  }

  private parentResource(): Resource {
    return this.currentScope.get('scopeLevelResource');
  }

  private applyCommonBehaviorFor(method: string, resources: string[], options: any, block: Block) {
    if (resources.length > 1) {
      resources.forEach(r => (this as any)[method](r, options, block));
      return true;
    }
    if (this.currentScope.isResourceScope()) {
      this.nested(() => {
        (this as any)[method](resources.pop(), options, block);
      });
      return true;
    }
  }

  private mergePathScope(parent: string, child: string) {
    return Mapper.normalizePath(`${parent || ''}/${child}`);
  }
}
