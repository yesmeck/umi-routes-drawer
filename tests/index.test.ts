import { draw } from '../src';

describe('drawer', () => {
  it('resources', () => {
    const routes = draw(({ resources }) => {
      resources('posts');
    });
    expect(routes).toMatchSnapshot();
  });

  it('resource', () => {
    const routes = draw(({ resource }) => {
      resource('profile');
    });
    expect(routes).toMatchSnapshot();
  });

  it('nested resources', () => {
    const routes = draw(({ resources }) => {
      resources('posts', () => {
        resources('comments');
      });
    });
    expect(routes).toMatchSnapshot();
  });

  it('nested collections', () => {
    const routes = draw(({ page, resources, collection }) => {
      resources('posts', () => {
        collection(() => {
          page('pulse');
        });
      });
    });
    expect(routes).toMatchSnapshot();
  });

  it('nested remember', () => {
    const routes = draw(({ page, resources, member }) => {
      resources('posts', () => {
        member(() => {
          page('draft');
        });
      });
    });
    expect(routes).toMatchSnapshot();
  });
});
