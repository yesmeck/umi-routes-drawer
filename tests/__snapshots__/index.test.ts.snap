// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawer nested collections 1`] = `
Array [
  Route {
    "name": "pulse_posts",
    "path": "/posts/pulse",
    "precedence": 0,
  },
  Route {
    "name": "posts",
    "path": "/posts",
    "precedence": 1,
  },
  Route {
    "name": "new_post",
    "path": "/posts/new",
    "precedence": 2,
  },
  Route {
    "name": "post_edit",
    "path": "/posts/:id/edit",
    "precedence": 3,
  },
  Route {
    "name": "post",
    "path": "/posts/:id",
    "precedence": 4,
  },
]
`;

exports[`drawer nested remember 1`] = `
Array [
  Route {
    "name": "post_draft",
    "path": "/posts/:id/draft",
    "precedence": 0,
  },
  Route {
    "name": "posts",
    "path": "/posts",
    "precedence": 1,
  },
  Route {
    "name": "new_post",
    "path": "/posts/new",
    "precedence": 2,
  },
  Route {
    "name": "post_edit",
    "path": "/posts/:id/edit",
    "precedence": 3,
  },
  Route {
    "name": "post",
    "path": "/posts/:id",
    "precedence": 4,
  },
]
`;

exports[`drawer nested resources 1`] = `
Array [
  Route {
    "name": "comments",
    "path": "/posts/:postId/comments",
    "precedence": 0,
  },
  Route {
    "name": "new_comment",
    "path": "/posts/:postId/comments/new",
    "precedence": 1,
  },
  Route {
    "name": "comment_edit",
    "path": "/posts/:postId/comments/:id/edit",
    "precedence": 2,
  },
  Route {
    "name": "comment",
    "path": "/posts/:postId/comments/:id",
    "precedence": 3,
  },
  Route {
    "name": "posts",
    "path": "/posts",
    "precedence": 4,
  },
  Route {
    "name": "new_post",
    "path": "/posts/new",
    "precedence": 5,
  },
  Route {
    "name": "post_edit",
    "path": "/posts/:id/edit",
    "precedence": 6,
  },
  Route {
    "name": "post",
    "path": "/posts/:id",
    "precedence": 7,
  },
]
`;

exports[`drawer resource 1`] = `
Array [
  Route {
    "name": "new_profile",
    "path": "/profile/new",
    "precedence": 0,
  },
  Route {
    "name": "profile_edit",
    "path": "/profile/edit",
    "precedence": 1,
  },
  Route {
    "name": "profile",
    "path": "/profile",
    "precedence": 2,
  },
]
`;

exports[`drawer resources 1`] = `
Array [
  Route {
    "name": "posts",
    "path": "/posts",
    "precedence": 0,
  },
  Route {
    "name": "new_post",
    "path": "/posts/new",
    "precedence": 1,
  },
  Route {
    "name": "post_edit",
    "path": "/posts/:id/edit",
    "precedence": 2,
  },
  Route {
    "name": "post",
    "path": "/posts/:id",
    "precedence": 3,
  },
]
`;
